### ISSUE TITLE
Global search and replace

### ISSUE THREAD

**Issue by Robinlovelace** [rstudio/rstudio#2066]
Another feature request (hence lack of system info), this one is for global search and replace functionality, perhaps implemented in the Window that pops-up when you press `Ctrl-Shift-F`. I opened this a while back on the old feature request site: https://support.rstudio.com/hc/en-us/community/posts/205333518-Global-search-and-replace

**Comment by ronblum** [comment]
Thanks @Robinlovelace. Just to spell it out—you'd like to do a replace (interactive or automatic) on all matching files in the directory tree?

**Comment by Robinlovelace** [comment]
Yes - interactive would be a bonus. Currently I use [regexxer](https://github.com/GNOME/regexxer) for such tasks, as illustrated below:

![peek 2018-01-26 14-28](https://user-images.githubusercontent.com/1825120/35444091-4c645f22-02a5-11e8-81a8-f9a5d020701c.gif)

**Comment by ronblum** [comment]
Thanks for the example! I've added this to the enhancements list to review for future releases.

**Comment by jennybc** [comment]
Related but perhaps should be separate issue: with the current "Find in Files" functionality, I get results in a new tab, in the same pane as Console, Terminal, etc. But this text cannot be copied to the clipboard, which can be frustrating if I'm, saying, working through a list of things I need to attend to. It is great that I can *click* on this text to go to specific loci and I imagine there is tension between that and the ability to copy.

**Comment by faridcher** [comment]
related to SO post: 
https://stackoverflow.com/questions/45894069/find-and-replace-text-in-all-files-rstudio/52096986#52096986

**Comment by billdenney** [comment]
I have the same issue relatively commonly.

I found the follow SO answer helpful for an R function that does the search and replace for you: https://stackoverflow.com/a/53397980/3831096

I'll also occasionally switch over to emacs for dired mode.

When implementing the function, it would help to show all the found items before the replace as is done for the current `ctrl-shift-f` command (optionally) so that the user could confirm that they will not make a mistake across many files.

**Comment by NanisTe** [comment]
Hi,

I think it would already help if the existing find and replace function would not always empty its entries when switching to another file.

Thanks for considering a solution for us.

**Comment by ju-rgen** [comment]
What is the problem in resolving this basic issue?

**Comment by Robinlovelace** [comment]
Heads-up @NanisTe I've also wondered that, and I imagine that issue would be quicker to fix. Assuming RStudio developers (thanks for awesome open source software guys!) agree, I would suggest opening a separate issue on that topic. 

@ju-rgen unsure what the problem is or if there is one, I imagine that this issue is not a priority but that it's climbing up the list based on the number of upvotes! Also I suspect it takes time to think over the option space of ways to implement this. I'd be interested to know how what the options are and how much work it would be, suspect that members of the community may be up for helping out with ideas/testing and maybe even code. Unfortunately I cannot program in languages in which RStudio is written. Any C++/Java devs out there know how to solve this? I'm not even sure which language the solution would be written in.

**Comment by jmcphers** [comment]
We have a beginner's guide to feature writing here: https://github.com/rstudio/rstudio/wiki/A-Beginner's-Guide-to-Writing-RStudio-IDE-Features

And the code for the existing find-in-files system (which would be extended to do find and replace) is here:

https://github.com/rstudio/rstudio/blob/master/src/cpp/session/modules/SessionFind.cpp

**Comment by Robinlovelace** [comment]
Nice, thanks for the link to the source code :+1:  Almost enough of an incentive to learn C++, on top of all the other ones.

**Comment by Robinlovelace** [comment]
Awesome work guys, many thanks @melissa-barca for programming this into RStudio, looking forward to testing it!

**Comment by ronblum** [comment]
@Robinlovelace Thank you for the kind note, and not only for testing it, but for suggesting it in the first place—it's clearly a popular request, with 43 👍s!

**Comment by melissa-barca** [comment]
Thank you @Robinlovelace I look forward to your feedback :)

**Comment by Robinlovelace** [comment]
On a related note I've found a command line solution which has become my go-to find and replace tool, the rust crate sd: https://github.com/chmln/sd has a relatively high barrier to entry. Having a solution built into RStudio will save lots of time for lots of people so thanks again, looking forward to giving it a spin.

**Comment by ronblum** [comment]
Verified in RStudio Desktop 1.3.743 on MacOS 10.15.3 and RStudio Server 1.3.730 on Red Hat 7.8.

**Comment by Robinlovelace** [comment]
Quick follow-up, how does one use it? Checking out Version 1.3.776 and trying Ctl-F and Ctl-Shift-F it's not immediately clear.

**Comment by Robinlovelace** [comment]
Update, apologies, it is clear. Ctl-Shift-F does it:

![image](https://user-images.githubusercontent.com/1825120/74183992-dd6f5b00-4c3d-11ea-9561-f9f595dfc343.png)

**Comment by Robinlovelace** [comment]
Feedback having tested it, it seems to work on the UI but according to my tests does not actually replace the files... Can share a screenshot of this. Any ideas @ronblum ?

**Comment by Robinlovelace** [comment]
![Peek 2020-02-19 10-46](https://user-images.githubusercontent.com/1825120/74827327-25107980-5305-11ea-822f-8ab9ae30ff7f.gif)

**Comment by melissa-barca** [comment]
Hi @Robinlovelace I just tested our daily build (https://dailies.rstudio.com/) on Ubuntu and was not able to reproduce this issue. Can you confirm if you have this issue with the most recent build and what OS Version you're using? Thank you.

**Comment by Robinlovelace** [comment]
Apologies, should have listed version, I'm on the latest preview version, Version 1.3.820.

**Comment by melissa-barca** [comment]
Thank you @Robinlovelace I am not able to reproduce this issue with the current preview either. Can you please confirm the OS version you're using?

**Comment by Robinlovelace** [comment]
Sorry @melissa-barca I tried it on the latest daily version and it worked. The behaviour shown in the screenshot above was on Ubuntu 18.04. I'm very happy with the performance of this feature on the daily version. One question though, how to replace items one-by-one? I'll create another GIF to show what I mean...

**Comment by Robinlovelace** [comment]
![Peek 2020-02-20 21-08](https://user-images.githubusercontent.com/1825120/74979161-f1794080-5425-11ea-89c5-8a59930a4543.gif)

In that example, it would be useful to have a button to replace one instance of 'geofabrik' at a time, not all instances want to be be replaced. Is that possible? Also it seemed something went wrong in that case, not sure what though and it definitely worked in a previous test. 

Hope this is useful and not a hassle for you, I really appreciate the work that has gone into it and it looks awesome very useful new feature.

**Comment by Robinlovelace** [comment]
Second attempt shown below, with a bit of a cryptic error message: 


![image](https://user-images.githubusercontent.com/1825120/74979675-02768180-5427-11ea-821a-df6bc3cb690d.png)

**Comment by melissa-barca** [comment]
Unfortunately at this time you cannot replace instances one at a time, but it's definitely an enhancement we'll consider for the future. The cryptic error message you're seeing is generated when we attempt to overwrite the original file with a file containing the replaced text. Is this something you're seeing regularly?

**Comment by Robinlovelace** [comment]
Thanks for the clarification. Is it worth opening a new issue to track ideas/development related to such an enhancement. I imagine the hard part is now done and that to add a new button 'Replace' (as opposed to 'Replace all') would increase the usefulness of this newfound functionality.

Regarding the error message, I've only tested the new functionality 3 times. The first time it worked fine, the second 2 times it failed.

Reproducible example starting from the command-line:

```bash
git clone git@github.com:ITSLeeds/geofabrik
rstudio geofabrik/geofabrik.Rproj
# then in RStudio try replacing all instances of 'geofabrik' with 'osmextractr' to rename it
```

**Comment by melissa-barca** [comment]
@Robinlovelace Yes, please open a new issue for this enhancement or I will open one. Unfortunately I was not able to reproduce the example you provided.

**Comment by Robinlovelace** [comment]
Job done, heads-up @melissa-barca and many thanks for great work on this awesome IDE.

**Comment by Robinlovelace** [comment]
Sorry to report this, but the 'Replace all' button is not seeming to update files on the HD on the latest daily version as far as I can tell, as documented in the more-or-less reproducible GIF below (from the itsleeds/pct github repo):

![Peek 2020-03-01 20-04](https://user-images.githubusercontent.com/1825120/75632875-188afb80-5bf8-11ea-9e82-dac93bebf2e2.gif)

**Comment by Robinlovelace** [comment]
On RStudio
Version 1.3.776
© 2009-2020 RStudio, Inc.
"Middlemist Red" (1d8a5728, 2020-01-23) for Unknown OS
Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) QtWebEngine/5.12.6 Chrome/69.0.3497.128 Safari/537.36

**Comment by melissa-barca** [comment]
Hi @Robinlovelace if you close and reopen the file does it contain the changes? We are tracking that the file doesn't always automatically refresh with the changes on issue #6237.

**Comment by Robinlovelace** [comment]
> if you close and reopen the file does it contain the changes?

Not on my computer. I do `git diff` after clicking 'Replace all' and nothing has changed on the HD it seems.

**Comment by melissa-barca** [comment]
@Robinlovelace are you able to provide an strace of the rsession from immediately before performing the replace to immediately after? Could you also run and provide a diagnostics report (https://support.rstudio.com/hc/en-us/articles/200321257-Running-a-Diagnostics-Report) after attempting a global replace? Thank you!

**Comment by Robinlovelace** [comment]
Sorry @melissa-barca it's a busy week so not sure I'll have time before next week.

**Comment by melissa-barca** [comment]
@Robinlovelace I understand, thanks again for your help looking into this!

**Comment by Robinlovelace** [comment]
I've generated a diagnostic report. All kinds of things in there about my set-up. How should I share it?

**Comment by melissa-barca** [comment]
Hey @Robinlovelace I really only need to see the Log file portion without any of your set up info. Could you share this portion here? Thank you.

**Comment by Robinlovelace** [comment]
Cannot see anything relevant in the log file but this shows in the find and replace tab, for example:

![image](https://user-images.githubusercontent.com/1825120/76034647-6a54be00-5f37-11ea-9bff-cd274b817be9.png)

**Comment by melissa-barca** [comment]
Hey @Robinlovelace this looks different from the original video you posted. When a replace fails, we expect the Error to be returned as it did in your most recent screenshot, highlighted in red -- though I need more information to debug this error. In your original video, the output was green as if the replace had occurred but then the files weren't updated. These are two different issues. Have you had a successful find/replace since your first round of testing?

**Comment by Robinlovelace** [comment]
> Have you had a successful find/replace since your first round of testing?

Afraid not, tested briefly on laptop and desktop although have gone back to using the released version of RStudio for now.

**Comment by ronblum** [comment]
@Robinlovelace and @melissa-barca I'm unable to reproduce the issue on Ubuntu 18.04.4—it works fine. Is that the correct OS? Also, I am still running into issue #6048 (though it works on MacOS).

**Comment by Robinlovelace** [comment]
Sorry to be the bearer of bad news but global replace just crashed my RStudio session. Screengrab below. Context I asked it to do a globaly replace in the robinlovelace/cyclestreets repo on Ubuntu 18.04, it seemed to hang for ~10 seconds before I hit 'Stop' and it crashed.

![image](https://user-images.githubusercontent.com/1825120/79991111-cfde1900-84a9-11ea-80b3-a6ef4a13b1f3.png)

**Comment by Robinlovelace** [comment]
On Version 1.3.938

**Comment by Robinlovelace** [comment]
I ended up using the command-line `sd` tool as follows to achieve the desired result:

```bash
sd 'cutoff = 20' 'cutoff = 50' $(fd R)
```

**Comment by Robinlovelace** [comment]
https://github.com/Robinlovelace/cyclestreets/commit/028ed1c3efb17dc8af6cf62762201a0ce484f467

**Comment by melissa-barca** [comment]
Hi @Robinlovelace I attempted to reproduce your issue on Ubuntu 18.04 but was able to perform the replace as expected. Are you able to consistently reproduce an example? If so could you please open a new issue with that description? Thanks!

**Comment by Robinlovelace** [comment]
I will open an issue if I can reproduce. Just a comment here for now as may be a one-off.

**Comment by AltfunsMA** [comment]
I have a similar issue. The first two images show output for a search and replace that failed completely. The errors reported are rather cryptic. The overall count of successes vs fails seems totally random: 64 vs 11, when actually nothing seems to have been replaced. Replacement didn't happen in the first file, either, although it reports no error.

When I tried a really small and simple test with just two fake files and one word to be replaced, the whole replacing got stuck. See third image below.

Running version 1.3.1073 on Ubuntu 18.04.4


1
![image](https://user-images.githubusercontent.com/26325512/92189398-e2c61880-eea1-11ea-9221-df826092b0cb.png)
2
![image](https://user-images.githubusercontent.com/26325512/92189427-f7a2ac00-eea1-11ea-9f05-6b10259a21af.png)
3
![image](https://user-images.githubusercontent.com/26325512/92189819-faea6780-eea2-11ea-91fe-4b0bfccb4fc3.png)
